### PR TITLE
Adds a flag to the clean after checkout and clean before checkout adv…

### PR DESCRIPTION
**Pull request by roguishmountain**
…anced options

See bug report https://issues.jenkins-ci.org/browse/JENKINS-26660 for more details

### CODE REVIEW

**Comment by roguishmountain** [comment]
Depends on https://github.com/jenkinsci/git-client-plugin/pull/222

**Comment by MarkEWaite** [comment]
I uploaded a build of required [git client plugin build](https://repo.jenkins-ci.org/snapshots/org/jenkins-ci/plugins/git-client/2.7.0-SNAPSHOT/git-client-2.7.0-20171216.041703-3.hpi) and the [git plugin build](https://repo.jenkins-ci.org/snapshots/org/jenkins-ci/plugins/git/3.7.0-SNAPSHOT/git-3.7.0-20171216.043204-1.hpi).  

The git client is based on the latest master branch.  The git plugin is the latest master branch plus a browser guesser enhancement and a fix for a performance problem due to excessive calls to rev-parse.

**Comment by leandro-lucarella-sociomantic** [comment]
For what is worth (and sorry for the delay), the tests also were well on our side, but we didn't test with pipelines.

**Comment by akoeplinger** [comment]
What's the current status of this PR?

**Comment by akoeplinger** [comment]
@MarkEWaite is the Pipeline work the only remaining issue before this PR can be merged?

**Comment by roguishmountain** [comment]
@MarkEWaite the most recent commit fixes the pipeline syntax suggestions.

File: src/main/resources/hudson/plugins/git/extensions/impl/CleanBeforeCheckout/config.groovy [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package hudson.plugins.git.extensions.impl.CleanBeforeCheckout;
+
+
+def f = namespace(lib.FormTagLib);
+
+f.entry(title:_("Clean submodules (Add additional -f flag)"), field:"cleanSubmodule") {
+    f.checkbox()
+}
```
Comment (darxriggs) [inline src/main/resources/hudson/plugins/git/extensions/impl/CleanBeforeCheckout/config.groovy]: A newline should be added here.

File: src/main/java/hudson/plugins/git/extensions/impl/CleanBeforeCheckout.java [hunk 2]
```diff
@@ -17,8 +17,19 @@
  * @author David S Wang
  */
 public class CleanBeforeCheckout extends GitSCMExtension {
+
+    private boolean cleanSubmodule;
     @DataBoundConstructor
+    public CleanBeforeCheckout(final boolean cleanSubmodule) {
```
Comment (darxriggs) [inline src/main/java/hudson/plugins/git/extensions/impl/CleanBeforeCheckout.java]: Please have a look at https://jenkins.io/doc/developer/plugin-development/pipeline-integration/#constructor-vs-setters. I suggest to not add another constructor (this procedure is hard to maintain when more and more parameters are added) but to use a setter method annotated with `@DataBoundSetter`.

File: src/main/resources/hudson/plugins/git/extensions/impl/CleanBeforeCheckout/config.groovy [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package hudson.plugins.git.extensions.impl.CleanBeforeCheckout;
```
Comment (darxriggs) [inline src/main/resources/hudson/plugins/git/extensions/impl/CleanBeforeCheckout/config.groovy]: Semicolons are optional in Groovy and it is common to not use them.

**Comment by leandro-lucarella-sociomantic** [comment]
Was this resolved or abandoned?

**Comment by MarkEWaite** [comment]
Neither resolved nor abandoned.  I've assigned it to the milestone "Later" and closed the pull request for now.  My priorities for development on the git plugin are:

1. Resolve open blocking issues in the [4.0 milestone](https://github.com/jenkinsci/git-plugin/milestone/3)
2. Complete git plugin 4.0 testing and release
3. Resolve BuildData bloat - [JENKINS-19022](https://issues.jenkins-ci.org/browse/JENKINS-19022) as described in "[BuildData](https://github.com/jenkinsci/git-plugin/pulls?q=label%3A%221+-+BuildData%22+is%3Aclosed)" label
4. Release git plugin version with BuildData bloat fix
5. Improve notifyCommit handling as described in [notifyCommit label](https://github.com/jenkinsci/git-plugin/pulls?q=label%3A%222+-+notifyCommit%22+is%3Aclosed)
6. Release git plugin version with notifyCommit improvements
7. Improve submodule handling as described in [submodules label](https://github.com/jenkinsci/git-plugin/pulls?q=label%3A%223+-+Submodules%22+is%3Aclosed)
8. Release git plugin version with submodule handling improvements
9. Improve repository caching in the plugin as described in the [caching label](https://github.com/jenkinsci/git-plugin/pulls?q=label%3A%224+-+Caching%22+is%3Aclosed)
10. Release git plugin version with caching improvements
11. Improve changelog handling based on pull requests in [changelog label](https://github.com/jenkinsci/git-plugin/pulls?q=label%3A%225+-+Changelog%22+is%3Aclosed)
12. Release git plugin version with changelog improvements

File: src/main/java/hudson/plugins/git/extensions/impl/CleanBeforeCheckout.java (hunk 2 above)
Comment (darxriggs) [reply src/main/java/hudson/plugins/git/extensions/impl/CleanBeforeCheckout.java]: @roguishmountain did you have some time to look at this?

File: src/main/java/hudson/plugins/git/extensions/impl/CleanCheckout.java [hunk 4]
```diff
@@ -17,8 +17,19 @@
  * @author Kohsuke Kawaguchi
  */
 public class CleanCheckout extends GitSCMExtension {
+
+    private boolean cleanSubmodule;
     @DataBoundConstructor
+    public CleanCheckout(final boolean cleanSubmodule) {
```
Comment (darxriggs) [inline src/main/java/hudson/plugins/git/extensions/impl/CleanCheckout.java]: See the [comment](https://github.com/jenkinsci/git-plugin/pull/449/files#r242635862) on `CleanBeforeCheckout`.

**Comment by darxriggs** [comment]
@roguishmountain can you have a look at the review comments?

**Comment by slonopotamus** [comment]
See #792

**Comment by MarkEWaite** [comment]
Merged into master as #792 .  Planned for release in git plugin 4.1.0

**Pull request closed without merging**